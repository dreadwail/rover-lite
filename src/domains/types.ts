import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
