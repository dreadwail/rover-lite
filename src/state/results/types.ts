export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
