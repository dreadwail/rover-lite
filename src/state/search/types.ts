import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
