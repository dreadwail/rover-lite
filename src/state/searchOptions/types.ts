export interface OptionCategory {
  readonly categoryId: string;
  readonly rank: number;
  readonly displayName: string;
}

export interface BasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly category: OptionCategory;
  readonly rank: number;
  readonly helpText?: string;
}

export interface DateOption extends BasicOption {
  readonly type: 'date';
}

export interface Choice {
  readonly value: any;
  readonly displayName: string;
}

export interface ChoiceOption extends BasicOption {
  readonly type: 'choice';
  readonly choices: Choice[];
}

export interface MultipleChoiceOption extends BasicOption {
  readonly type: 'multiple choice';
  readonly choices: Choice[];
}

export interface IntegerOption extends BasicOption {
  readonly type: 'integer';
  readonly minValue: number;
  readonly maxValue: number;
}

export interface BooleanOption extends BasicOption {
  readonly type: 'boolean';
}

export type Option =
  | BasicOption
  | DateOption
  | ChoiceOption
  | MultipleChoiceOption
  | IntegerOption
  | BooleanOption;

export const isBooleanOption = (option: Option): option is BooleanOption =>
  option.type === 'boolean';
export const isIntegerOption = (option: Option): option is IntegerOption =>
  option.type === 'integer';
export const isMultipleChoiceOption = (option: Option): option is MultipleChoiceOption =>
  option.type === 'multiple choice';
export const isChoiceOption = (option: Option): option is ChoiceOption => option.type === 'choice';
export const isDateOption = (option: Option): option is DateOption => option.type === 'date';
export const isBasicOption = (option: Option): option is BasicOption =>
  [isBooleanOption, isIntegerOption, isChoiceOption, isMultipleChoiceOption, isDateOption].every(
    check => !check(option)
  );

export type Options = Record<string, Option>;
export type OptionsByCategoryId = Record<string, Options>;

export interface SearchOptions {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly options: Options;
}
