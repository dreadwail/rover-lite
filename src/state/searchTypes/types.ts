export type SearchType =
  | 'overnight-boarding'
  | 'overnight-traveling'
  | 'drop-in'
  | 'doggy-day-care'
  | 'dog-walking';

export type Endpoints = { [T in SearchType]?: string };

export interface SearchTypes {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly endpoints: Endpoints;
}
