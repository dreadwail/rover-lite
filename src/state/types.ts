import { Results } from './results/types';
import { Search } from './search/types';
import { SearchOptions } from './searchOptions/types';
import { SearchTypes } from './searchTypes/types';

export interface RoverSearchState {
  readonly searchTypes: SearchTypes;
  readonly searchOptions: SearchOptions;
  readonly search: Search;
  readonly results: Results;
}
