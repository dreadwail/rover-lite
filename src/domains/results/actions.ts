import { Action } from 'redux';

import { request } from '../../api/http';
import { urlEncodeQuery } from '../../api/query';
import { Listing } from '../../state/results/types';
import { SearchInputs } from '../../state/search/types';
import debouncedAction from '../debouncedThunk';
import { getInputs, getLatLong, getSearchType } from '../search/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, RawSearchResponse } from './normalize';

export interface FetchResultsPendingAction extends Action {
  readonly type: 'FETCH_RESULTS.PENDING';
}

export interface FetchResultsFulfilledAction extends Action {
  readonly type: 'FETCH_RESULTS.FULFILLED';
  readonly payload: Listing[];
}

export interface FetchResultsRejectedAction extends Action {
  readonly type: 'FETCH_RESULTS.REJECTED';
  readonly payload: {
    readonly error: any;
  };
  readonly error: true;
}

const fetchResultsPending = (): FetchResultsPendingAction => ({
  type: 'FETCH_RESULTS.PENDING',
});

const fetchResultsFulfilled = (listings: Listing[]): FetchResultsFulfilledAction => ({
  type: 'FETCH_RESULTS.FULFILLED',
  payload: listings,
});

const fetchResultsRejected = (error: any): FetchResultsRejectedAction => ({
  type: 'FETCH_RESULTS.REJECTED',
  payload: {
    error,
  },
  error: true,
});

export type FetchResults = () => PromiseThunkAction<void>;
export const fetchResults: FetchResults = () => async (dispatch, getState) => {
  dispatch(fetchResultsPending());
  try {
    const state = getState();

    const searchType = getSearchType(state);
    const [latitude, longitude] = getLatLong(state);

    const defaultInputs: SearchInputs = {
      centerlat: latitude,
      centerlng: longitude,
      service_type: searchType,
    };
    const userInputs = getInputs(state);
    const inputs = { ...defaultInputs, ...userInputs };
    const encodedQuery = urlEncodeQuery(inputs);

    const urlBase = 'https://www.rover.com/api/v3/search';
    const url = `${urlBase}/${searchType}/?${encodedQuery}`;

    const response = await request<RawSearchResponse>('GET', url);

    const normalizedListings = normalize(response);
    dispatch(fetchResultsFulfilled(normalizedListings));
  } catch (err) {
    dispatch(fetchResultsRejected(err));
  }
};

export const debouncedFetchResults = debouncedAction(fetchResults, 250);
