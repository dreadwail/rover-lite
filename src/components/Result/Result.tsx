import React, { FC } from 'react';

import { Listing } from '../../state/results/types';
import ExternalLink from '../ExternalLink';

export interface ResultProps {
  readonly listing: Listing;
}

const Result: FC<ResultProps> = ({ listing }) => {
  const {
    thumbnailUrl,
    name,
    rank,
    webUrl,
    tagline,
    neighborhood,
    city,
    state,
    zip,
    repeatClientCount,
    reviewCount,
    featuredReview,
    currencySymbol,
    price,
    priceUnit,
  } = listing;

  const locationParts = [neighborhood, city, state].filter(part => !!part);
  const location = `${locationParts.join(', ')} ${zip}`;

  return (
    <div className="row">
      <div className="col">
        <img src={thumbnailUrl} alt={name} className="img-thumbnail img-fluid mx-auto d-block" />
      </div>
      <div className="col-sm-7 col-md-8">
        <div className="text-center text-sm-left">
          <p className="lead mb-0">
            {rank}. <ExternalLink href={webUrl}>{name}</ExternalLink>
          </p>
          <p className="mb-0">{tagline}</p>
          <p className="mb-0">{location}</p>
          <div className="my-1">
            <a href="/" className="badge badge-light border mr-1 text-uppercase">
              {repeatClientCount} repeat clients
            </a>
            <a href="/" className="badge badge-light border mr-1 text-uppercase">
              {reviewCount} reviews
            </a>
          </div>
          <p className="text-muted">
            <em>"{featuredReview}"</em>
          </p>
        </div>
      </div>
      <div className="col-sm-2">
        <div className="d-flex flex-column text-center">
          <span>from</span>
          <span className="lead text-danger font-weight-bold">
            {currencySymbol}
            {price}
          </span>
          <span>per {priceUnit}</span>
        </div>
      </div>
    </div>
  );
};

export default Result;
