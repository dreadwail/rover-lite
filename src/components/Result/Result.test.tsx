import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import { Listing } from '../../state/results/types';

import Result from './Result';

describe('Result', () => {
  let rendered: ShallowWrapper;

  beforeEach(() => {
    const listing: Listing = {
      thumbnailUrl: 'https://example.com/profile.png',
      rank: 0,
      name: 'Test Name',
      badges: [
        {
          content: 'This sitter is Certified as Awesome.',
          imageUrl: 'https://example.com/badge.png',
          title: 'Certified Awesome',
        },
      ],
      tagline: 'Enthusiastic dog lover in Ballard.',
      neighborhood: 'Ballard',
      city: 'Seattle',
      state: 'WA',
      zip: '98117',
      repeatClientCount: 42,
      reviewCount: 12,
      ratingAverage: 4.8,
      featuredReview: 'This sitter is awesome.',
      currencySymbol: '$',
      price: 42.12,
      priceUnit: 'night',
      webUrl: 'https://www.rover.com/members/not-a-real-profile-just-a-test/',
    };
    rendered = shallow(<Result listing={listing} />);
  });

  it('renders correctly', () => {
    expect(rendered).toMatchSnapshot();
  });
});
