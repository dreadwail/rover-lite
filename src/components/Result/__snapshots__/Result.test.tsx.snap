// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result renders correctly 1`] = `
<div
  className="row"
>
  <div
    className="col"
  >
    <img
      alt="Test Name"
      className="img-thumbnail img-fluid mx-auto d-block"
      src="https://example.com/profile.png"
    />
  </div>
  <div
    className="col-sm-7 col-md-8"
  >
    <div
      className="text-center text-sm-left"
    >
      <p
        className="lead mb-0"
      >
        0
        . 
        <ExternalLink
          href="https://www.rover.com/members/not-a-real-profile-just-a-test/"
        >
          Test Name
        </ExternalLink>
      </p>
      <p
        className="mb-0"
      >
        Enthusiastic dog lover in Ballard.
      </p>
      <p
        className="mb-0"
      >
        Ballard, Seattle, WA 98117
      </p>
      <div
        className="my-1"
      >
        <a
          className="badge badge-light border mr-1 text-uppercase"
          href="/"
        >
          42
           repeat clients
        </a>
        <a
          className="badge badge-light border mr-1 text-uppercase"
          href="/"
        >
          12
           reviews
        </a>
      </div>
      <p
        className="text-muted"
      >
        <em>
          "
          This sitter is awesome.
          "
        </em>
      </p>
    </div>
  </div>
  <div
    className="col-sm-2"
  >
    <div
      className="d-flex flex-column text-center"
    >
      <span>
        from
      </span>
      <span
        className="lead text-danger font-weight-bold"
      >
        $
        42.12
      </span>
      <span>
        per 
        night
      </span>
    </div>
  </div>
</div>
`;
